File: frontend/src/services/api/problem.ts
import { apiClient } from './client';

export type Platform = 'leetcode' | 'gfg' | 'hackerrank' | 'codeforces' | 'interviewbit' | 'other';
export type Difficulty = 'easy' | 'medium' | 'hard';
export type Status = 'solved' | 'revisit' | 'attempted' | 'skipped';

export interface Problem {
  id: string;
  title: string;
  platform: Platform;
  topic?: string;
  difficulty: Difficulty;
  status: Status;
  platform_url?: string;
  notes?: string;
  is_bookmarked: boolean;
  user_id: string;
  created_at: string;
  updated_at: string;
}

export type ProblemCreate = Omit<Problem, 'id' | 'user_id' | 'created_at' | 'updated_at'>;
export type ProblemUpdate = Partial<ProblemCreate>;

export interface ProblemQueryParams {
  search?: string;
  difficulty?: Difficulty;
  status?: Status;
  platform?: string;
  is_bookmarked?: boolean;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  skip?: number;
  limit?: number;
}

export const getProblems = async (params?: ProblemQueryParams): Promise<Problem[]> => {
  const { data } = await apiClient.get('/problems/', { params });
  return data?.data ?? data;
};

export const getProblem = async (id: string): Promise<Problem> => {
  const { data } = await apiClient.get(`/problems/${id}`);
  return data?.data ?? data;
};

export const createProblem = async (problem: ProblemCreate): Promise<Problem> => {
  const { data } = await apiClient.post('/problems/', problem);
  return data?.data ?? data;
};

export const updateProblem = async (id: string, problem: ProblemUpdate): Promise<Problem> => {
  const { data } = await apiClient.put(`/problems/${id}`, problem);
  return data?.data ?? data;
};

export const deleteProblem = async (id: string): Promise<void> => {
  await apiClient.delete(`/problems/${id}`);
};

export const bulkDeleteProblems = async (ids: string[]): Promise<void> => {
  await apiClient.post('/problems/bulk-delete', ids);
};

File: backend/app/schemas/problem.py
from typing import Optional, Literal
from pydantic import BaseModel, ConfigDict, UUID4
from datetime import datetime

class ProblemBase(BaseModel):
    title: str
    platform: str
    platform_url: Optional[str] = None
    topic: str
    difficulty: Literal['easy', 'medium', 'hard']
    status: Literal['solved', 'revisit', 'attempted', 'skipped'] = "solved"
    is_bookmarked: bool = False
    notes: Optional[str] = None
    time_complexity: Optional[str] = None
    space_complexity: Optional[str] = None

class ProblemCreate(ProblemBase):
    pass

class ProblemUpdate(ProblemBase):
    title: Optional[str] = None
    platform: Optional[str] = None
    topic: Optional[str] = None
    difficulty: Optional[Literal['easy', 'medium', 'hard']] = None
    status: Optional[Literal['solved', 'revisit', 'attempted', 'skipped']] = None
    is_bookmarked: Optional[bool] = None

class ProblemResponse(ProblemBase):
    id: UUID4
    user_id: UUID4
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)
